File: ios-native/VoiceAssistant.swift
import Foundation
import Speech
import Intents
import NaturalLanguage

// MARK: - Voice Assistant (Erica/Eno-like)

class VoiceAssistant: NSObject {
    static let shared = VoiceAssistant()
    
    private let speechRecognizer = SFSpeechRecognizer(locale: Locale(identifier: "en-US"))
    private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
    private var recognitionTask: SFSpeechRecognitionTask?
    private let audioEngine = AVAudioEngine()
    
    enum Command {
        case checkBalance
        case sendMoney(recipient: String, amount: Double)
        case viewSpending(period: String)
        case buyStock(symbol: String, shares: Int)
        case payBill(billType: String)
        case unknown(String)
    }
    
    // MARK: - Start Listening
    
    func startListening(completion: @escaping (Result<Command, Error>) -> Void) throws {
        // Cancel previous task
        recognitionTask?.cancel()
        recognitionTask = nil
        
        // Configure audio session
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
        try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
        
        recognitionRequest = SFSpeechAudioBufferRecognitionRequest()
        
        guard let recognitionRequest = recognitionRequest else {
            throw NSError(domain: "VoiceAssistant", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unable to create recognition request"])
        }
        
        recognitionRequest.shouldReportPartialResults = true
        
        // Configure microphone input
        let inputNode = audioEngine.inputNode
        let recordingFormat = inputNode.outputFormat(forBus: 0)
        
        inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { buffer, _ in
            recognitionRequest.append(buffer)
        }
        
        audioEngine.prepare()
        try audioEngine.start()
        
        // Start recognition
        recognitionTask = speechRecognizer?.recognitionTask(with: recognitionRequest) { [weak self] result, error in
            if let result = result {
                let transcript = result.bestTranscription.formattedString
                
                if result.isFinal {
                    self?.stopListening()
                    let command = self?.parseCommand(from: transcript) ?? .unknown(transcript)
                    completion(.success(command))
                }
            }
            
            if let error = error {
                self?.stopListening()
                completion(.failure(error))
            }
        }
    }
    
    func stopListening() {
        audioEngine.stop()
        audioEngine.inputNode.removeTap(onBus: 0)
        recognitionRequest?.endAudio()
        recognitionRequest = nil
        recognitionTask = nil
    }
    
    // MARK: - Command Parsing
    
    private func parseCommand(from transcript: String) -> Command {
        let lowercased = transcript.lowercased()
        
        // Check balance
        if lowercased.contains("balance") || lowercased.contains("how much") {
            return .checkBalance
        }
        
        // Send money
        if lowercased.contains("send") || lowercased.contains("transfer") || lowercased.contains("pay") {
            if let amount = extractAmount(from: lowercased),
               let recipient = extractRecipient(from: lowercased) {
                return .sendMoney(recipient: recipient, amount: amount)
            }
        }
        
        // View spending
        if lowercased.contains("spending") || lowercased.contains("spent") {
            let period = extractPeriod(from: lowercased)
            return .viewSpending(period: period)
        }
        
        // Buy stock
        if lowercased.contains("buy") && (lowercased.contains("share") || lowercased.contains("stock")) {
            if let shares = extractShares(from: lowercased),
               let symbol = extractStockSymbol(from: lowercased) {
                return .buyStock(symbol: symbol, shares: shares)
            }
        }
        
        // Pay bill
        if lowercased.contains("bill") {
            if let billType = extractBillType(from: lowercased) {
                return .payBill(billType: billType)
            }
        }
        
        return .unknown(transcript)
    }
    
    // MARK: - Natural Language Processing
    
    private func extractAmount(from text: String) -> Double? {
        let tagger = NLTagger(tagSchemes: [.lexicalClass])
        tagger.string = text
        
        var amount: Double?
        
        // Look for number words
        let numberWords = ["one": 1.0, "two": 2.0, "three": 3.0, "four": 4.0, "five": 5.0,
                          "ten": 10.0, "twenty": 20.0, "thirty": 30.0, "forty": 40.0, "fifty": 50.0,
                          "hundred": 100.0, "thousand": 1000.0]
        
        for (word, value) in numberWords {
            if text.contains(word) {
                amount = value
                break
            }
        }
        
        // Look for numeric values
        let pattern = "\\d+(\\.\\d+)?"
        if let regex = try? NSRegularExpression(pattern: pattern) {
            let matches = regex.matches(in: text, range: NSRange(text.startIndex..., in: text))
            if let match = matches.first {
                let range = Range(match.range, in: text)!
                amount = Double(text[range])
            }
        }
        
        return amount
    }
    
    private func extractRecipient(from text: String) -> String? {
        let tagger = NLTagger(tagSchemes: [.nameType])
        tagger.string = text
        
        var recipient: String?
        
        tagger.enumerateTags(in: text.startIndex..<text.endIndex, unit: .word, scheme: .nameType) { tag, range in
            if tag == .personalName {
                recipient = String(text[range])
                return false
            }
            return true
        }
        
        return recipient
    }
    
    private func extractPeriod(from text: String) -> String {
        if text.contains("today") {
            return "today"
        } else if text.contains("week") {
            return "week"
        } else if text.contains("month") {
            return "month"
        } else if text.contains("year") {
            return "year"
        }
        return "month"
    }
    
    private func extractShares(from text: String) -> Int? {
        let pattern = "(\\d+)\\s+(share|stock)"
        if let regex = try? NSRegularExpression(pattern: pattern) {
            let matches = regex.matches(in: text, range: NSRange(text.startIndex..., in: text))
            if let match = matches.first, match.numberOfRanges > 1 {
                let range = Range(match.range(at: 1), in: text)!
                return Int(text[range])
            }
        }
        return nil
    }
    
    private func extractStockSymbol(from text: String) -> String? {
        // Common stock names
        let stocks = ["apple": "AAPL", "microsoft": "MSFT", "google": "GOOGL", 
                     "amazon": "AMZN", "tesla": "TSLA", "meta": "META"]
        
        for (name, symbol) in stocks {
            if text.contains(name) {
                return symbol
            }
        }
        
        return nil
    }
    
    private func extractBillType(from text: String) -> String? {
        if text.contains("electric") {
            return "electricity"
        } else if text.contains("water") {
            return "water"
        } else if text.contains("internet") || text.contains("wifi") {
            return "internet"
        } else if text.contains("phone") {
            return "phone"
        }
        return nil
    }
    
    // MARK: - Execute Command
    
    func executeCommand(_ command: Command, completion: @escaping (Result<String, Error>) -> Void) {
        switch command {
        case .checkBalance:
            // Fetch balance
            let balance = "₦125,450.00"
            completion(.success("Your current balance is \(balance)"))
            
        case .sendMoney(let recipient, let amount):
            // Execute transfer
            completion(.success("Sending ₦\(amount) to \(recipient)"))
            
        case .viewSpending(let period):
            // Fetch spending
            let spending = "₦45,000"
            completion(.success("You spent \(spending) this \(period)"))
            
        case .buyStock(let symbol, let shares):
            // Execute stock purchase
            completion(.success("Buying \(shares) shares of \(symbol)"))
            
        case .payBill(let billType):
            // Pay bill
            completion(.success("Paying your \(billType) bill"))
            
        case .unknown(let text):
            completion(.success("I didn't understand: \(text)"))
        }
    }
}

// MARK: - Siri Shortcuts Integration

class SiriShortcutsManager {
    static let shared = SiriShortcutsManager()
    
    func donateCheckBalanceIntent() {
        let intent = CheckBalanceIntent()
        intent.suggestedInvocationPhrase = "Check my balance"
        
        let interaction = INInteraction(intent: intent, response: nil)
        interaction.donate { error in
            if let error = error {
                print("Failed to donate intent: \(error)")
            }
        }
    }
    
    func donateSendMoneyIntent(recipient: String, amount: Double) {
        let intent = SendMoneyIntent()
        intent.recipient = recipient
        intent.amount = NSNumber(value: amount)
        intent.suggestedInvocationPhrase = "Send money to \(recipient)"
        
        let interaction = INInteraction(intent: intent, response: nil)
        interaction.donate { error in
            if let error = error {
                print("Failed to donate intent: \(error)")
            }
        }
    }
}

// MARK: - Intent Definitions

class CheckBalanceIntent: INIntent {}
class SendMoneyIntent: INIntent {
    @NSManaged var recipient: String?
    @NSManaged var amount: NSNumber?
}

File: android-native/VoiceAssistant.kt
package com.pos54link.app.voice

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import kotlinx.coroutines.*

// Voice Assistant (Google Assistant Integration)

class VoiceAssistant(private val context: Context) {
    
    private var speechRecognizer: SpeechRecognizer? = null
    private var isListening = false
    
    sealed class Command {
        object CheckBalance : Command()
        data class SendMoney(val recipient: String, val amount: Double) : Command()
        data class ViewSpending(val period: String) : Command()
        data class BuyStock(val symbol: String, val shares: Int) : Command()
        data class PayBill(val billType: String) : Command()
        data class Unknown(val text: String) : Command()
    }
    
    fun startListening(callback: (Result<Command>) -> Unit) {
        if (isListening) return
        
        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context)
        
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
            putExtra(RecognizerIntent.EXTRA_LANGUAGE, "en-US")
            putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true)
        }
        
        speechRecognizer?.setRecognitionListener(object : RecognitionListener {
            override fun onReadyForSpeech(params: Bundle?) {
                isListening = true
            }
            
            override fun onBeginningOfSpeech() {}
            
            override fun onRmsChanged(rmsdB: Float) {}
            
            override fun onBufferReceived(buffer: ByteArray?) {}
            
            override fun onEndOfSpeech() {
                isListening = false
            }
            
            override fun onError(error: Int) {
                isListening = false
                callback(Result.failure(Exception("Speech recognition error: $error")))
            }
            
            override fun onResults(results: Bundle?) {
                isListening = false
                val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                if (matches != null && matches.isNotEmpty()) {
                    val transcript = matches[0]
                    val command = parseCommand(transcript)
                    callback(Result.success(command))
                }
            }
            
            override fun onPartialResults(partialResults: Bundle?) {}
            
            override fun onEvent(eventType: Int, params: Bundle?) {}
        })
        
        speechRecognizer?.startListening(intent)
    }
    
    fun stopListening() {
        speechRecognizer?.stopListening()
        speechRecognizer?.destroy()
        speechRecognizer = null
        isListening = false
    }
    
    private fun parseCommand(transcript: String): Command {
        val lowercased = transcript.lowercase()
        
        // Check balance
        if (lowercased.contains("balance") || lowercased.contains("how much")) {
            return Command.CheckBalance
        }
        
        // Send money
        if (lowercased.contains("send") || lowercased.contains("transfer") || lowercased.contains("pay")) {
            val amount = extractAmount(lowercased)
            val recipient = extractRecipient(lowercased)
            if (amount != null && recipient != null) {
                return Command.SendMoney(recipient, amount)
            }
        }
        
        // View spending
        if (lowercased.contains("spending") || lowercased.contains("spent")) {
            val period = extractPeriod(lowercased)
            return Command.ViewSpending(period)
        }
        
        // Buy stock
        if (lowercased.contains("buy") && (lowercased.contains("share") || lowercased.contains("stock"))) {
            val shares = extractShares(lowercased)
            val symbol = extractStockSymbol(lowercased)
            if (shares != null && symbol != null) {
                return Command.BuyStock(symbol, shares)
            }
        }
        
        // Pay bill
        if (lowercased.contains("bill")) {
            val billType = extractBillType(lowercased)
            if (billType != null) {
                return Command.PayBill(billType)
            }
        }
        
        return Command.Unknown(transcript)
    }
    
    private fun extractAmount(text: String): Double? {
        // Number words
        val numberWords = mapOf(
            "one" to 1.0, "two" to 2.0, "three" to 3.0, "four" to 4.0, "five" to 5.0,
            "ten" to 10.0, "twenty" to 20.0, "thirty" to 30.0, "forty" to 40.0, "fifty" to 50.0,
            "hundred" to 100.0, "thousand" to 1000.0
        )
        
        for ((word, value) in numberWords) {
            if (text.contains(word)) {
                return value
            }
        }
        
        // Numeric values
        val regex = Regex("\\d+(\\.\\d+)?")
        val match = regex.find(text)
        return match?.value?.toDoubleOrNull()
    }
    
    private fun extractRecipient(text: String): String? {
        // Look for names after "to"
        val toIndex = text.indexOf(" to ")
        if (toIndex != -1) {
            val afterTo = text.substring(toIndex + 4)
            val words = afterTo.split(" ")
            if (words.isNotEmpty()) {
                return words[0].capitalize()
            }
        }
        return null
    }
    
    private fun extractPeriod(text: String): String {
        return when {
            text.contains("today") -> "today"
            text.contains("week") -> "week"
            text.contains("month") -> "month"
            text.contains("year") -> "year"
            else -> "month"
        }
    }
    
    private fun extractShares(text: String): Int? {
        val regex = Regex("(\\d+)\\s+(share|stock)")
        val match = regex.find(text)
        return match?.groupValues?.get(1)?.toIntOrNull()
    }
    
    private fun extractStockSymbol(text: String): String? {
        val stocks = mapOf(
            "apple" to "AAPL", "microsoft" to "MSFT", "google" to "GOOGL",
            "amazon" to "AMZN", "tesla" to "TSLA", "meta" to "META"
        )
        
        for ((name, symbol) in stocks) {
            if (text.contains(name)) {
                return symbol
            }
        }
        return null
    }
    
    private fun extractBillType(text: String): String? {
        return when {
            text.contains("electric") -> "electricity"
            text.contains("water") -> "water"
            text.contains("internet") || text.contains("wifi") -> "internet"
            text.contains("phone") -> "phone"
            else -> null
        }
    }
    
    suspend fun executeCommand(command: Command): String {
        return when (command) {
            is Command.CheckBalance -> {
                "Your current balance is ₦125,450.00"
            }
            is Command.SendMoney -> {
                "Sending ₦${command.amount} to ${command.recipient}"
            }
            is Command.ViewSpending -> {
                "You spent ₦45,000 this ${command.period}"
            }
            is Command.BuyStock -> {
                "Buying ${command.shares} shares of ${command.symbol}"
            }
            is Command.PayBill -> {
                "Paying your ${command.billType} bill"
            }
            is Command.Unknown -> {
                "I didn't understand: ${command.text}"
            }
        }
    }
}

// Google Assistant Actions Integration

class GoogleAssistantManager(private val context: Context) {
    
    fun registerActions() {
        // Register app actions for Google Assistant
        // This would be configured in actions.xml
    }
    
    fun handleAssistantIntent(intent: Intent): String? {
        return when (intent.action) {
            "com.pos54link.CHECK_BALANCE" -> handleCheckBalance()
            "com.pos54link.SEND_MONEY" -> handleSendMoney(intent)
            "com.pos54link.VIEW_SPENDING" -> handleViewSpending(intent)
            else -> null
        }
    }
    
    private fun handleCheckBalance(): String {
        return "Your balance is ₦125,450.00"
    }
    
    private fun handleSendMoney(intent: Intent): String {
        val recipient = intent.getStringExtra("recipient") ?: "unknown"
        val amount = intent.getDoubleExtra("amount", 0.0)
        return "Sending ₦$amount to $recipient"
    }
    
    private fun handleViewSpending(intent: Intent): String {
        val period = intent.getStringExtra("period") ?: "month"
        return "You spent ₦45,000 this $period"
    }
}
